homeassistant/helpers/entityfilter.py

```diff
@@ -30,31 +39,110 @@ def convert_filter(config: Dict[str, List[str]]) -> Callable[[str], bool]:
+    exclude = config.get(CONF_EXCLUDE, {})
+    filt = generate_filter(
+        include.get(CONF_DOMAINS, []),
+        include.get(CONF_ENTITIES, []),
+        exclude.get(CONF_DOMAINS, []),
+        exclude.get(CONF_ENTITIES, []),
+        include.get(CONF_ENTITY_GLOBS, []),
+        exclude.get(CONF_ENTITY_GLOBS, []),
+    )
+    total_filters = sum(len(val) for val in include.values()) + sum(
+        len(val) for val in exclude.values()
+    )
+    setattr(filt, "config", config)
+    setattr(filt, "empty_filter", total_filters == 0)
+    return filt
+
+
+ALT_FILTER_SCHEMA_INNER = vol.Schema(
+    {
+        vol.Optional(CONF_DOMAINS, default=[]): vol.All(cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_ENTITY_GLOBS, default=[]): vol.All(
+            cv.ensure_list, [cv.string]
+        ),
+        vol.Optional(CONF_ENTITIES, default=[]): cv.entity_ids,
+    }
+)
+
+ALT_BASE_FILTER_SCHEMA = vol.Schema(
+    {
+        vol.Optional(
+            CONF_INCLUDE, default=ALT_FILTER_SCHEMA_INNER({})
+        ): ALT_FILTER_SCHEMA_INNER,
+        vol.Optional(
+            CONF_EXCLUDE, default=ALT_FILTER_SCHEMA_INNER({})
+        ): ALT_FILTER_SCHEMA_INNER,
+    }
+)
+
+ALT_FILTER_SCHEMA = vol.All(ALT_BASE_FILTER_SCHEMA, convert_filter_alt)
+
+
+def _glob_to_re(glob: str) -> Pattern:
+    """Translate and compile glob string into pattern."""
+    return re.compile(fnmatch.translate(glob))
+
+
+# It's safe since we don't modify it. And None causes typing warnings
+# pylint: disable=dangerous-default-value
 def generate_filter(
     include_domains: List[str],
     include_entities: List[str],
     exclude_domains: List[str],
     exclude_entities: List[str],
+    include_entity_globs: List[str] = [],
+    exclude_entity_globs: List[str] = [],
 ) -> Callable[[str], bool]:
     """Return a function that will filter entities based on the args."""
     include_d = set(include_domains)
     include_e = set(include_entities)
     exclude_d = set(exclude_domains)
     exclude_e = set(exclude_entities)
-
-    have_exclude = bool(exclude_e or exclude_d)
-    have_include = bool(include_e or include_d)
+    include_eg = set(map(_glob_to_re, include_entity_globs))
+    exclude_eg = set(map(_glob_to_re, exclude_entity_globs))
+
+    have_exclude = bool(exclude_e or exclude_d or exclude_eg)
+    have_include = bool(include_e or include_d or include_eg)
+
+    def entity_included(domain: str, entity_id: str) -> bool:
+        """Return true if entity matches inclusion filters."""
+        return (
+            entity_id in include_e
+            or domain in include_d
+            or any(glob_re.match(entity_id) for glob_re in include_eg)
+        )
+
+    def entity_excluded(domain: str, entity_id: str) -> bool:
+        """Return true if entity matches exclusion filters."""
+        return (
+            entity_id in exclude_e
+            or domain in exclude_d
+            or any(glob_re.match(entity_id) for glob_re in exclude_eg)
+        )
```

Comment:
This is going to be expensive, especially with a lot of regexes.

Reply:
Can you build a single include and exclude regex here instead?

Reply:
So that was why I always tried to put that test last to ensure that it would kick out of the conditional before it got there whenever possible. Also if a user isn't using the feature then `incude_eg`\`exclude_eg` will just be an empty array so that won't loop over anything right? Shouldn't be any regex compilation or match checking in that case to pay for I thought.

As for the single regex, I actually implemented it that way initially. But then before going with it I did a quick performance test comparing which is faster, comparing 100 smaller regexes against an entity ID or comparing 1 regex with 100 OR's against an entity ID. I did a 100 random GLOBs and tested against an ID that intentionally didn't match any so it went through all.

I found using a single regex was a lot slower.  Here's the output from my little test:
```
One RE - 0:00:00.012097
Many REs - 0:00:00.000005
```
Here's the code if you want to try it out for yourself: https://paste.ubuntu.com/p/ysj5Py6j29/ 

Anyway yea that was enough for me to adjust the implementation. And while you're definitely right about additional complexity and overhead here I feel like the current implementation just puts far too much burden on users to micromanage. You may have seen the recent community guide pop-up for [extending the life of your sd card](https://community.home-assistant.io/t/how-to-reduce-your-database-size-and-extend-the-life-of-your-sd-card/205299). It's essentially a consolidation of a number of other posts about it. People are listing out so many ids they have to use scripts to build the list and then manage the process in excel. It's quite overwhelming and results in a really long list of IDs to check for equality against. My thinking is that for many users the addition of GLOB support would result in a much smaller list and much less user burden here.

Reply:
Please use `timeit` for the benchmark https://docs.python.org/3/library/timeit.html

Also you should only compile the regex one time.  The current benchmark is doing it every time so its not an accurate approximation of what would actually happen.

Additionally the list of regexes at the top of the file should be real use cases to get a more realistic benchmark.

Reply:
@bdraco ok for the 500k entities, I'll get some numbers. I'll do like 100 of each option and throw 500k entities at it which don't match anything. I'll also get a baseline to see what overhead is added just by adding glob support even if users don't use it. Anything else?

Also are there any performance tests around filtering, recorder or logbook right now I can run and add to? I wasn't able to find any.

Reply:
> @bdraco ok for the 500k entities, I'll get some numbers. I'll do like 100 of each option and throw 500k entities at it which don't match anything. I'll also get a baseline to see what overhead is added just by adding glob support even if users don't use it. Anything else?

Probably best to assume 50% match and 50% don't match so we get a real world use case.
> 
> Also are there any performance tests around filtering, recorder or logbook right now I can run and add to? I wasn't able to find any.

I've fixed the benchmarking for logbook in this PR:

https://github.com/home-assistant/core/pull/36958#issuecomment-647021120

Reply:
> Please use `timeit` for the benchmark https://docs.python.org/3/library/timeit.html
Sorry wasn't aware of it, will do

> Also you should only compile the regex one time. The current benchmark is doing it every time so its not an accurate approximation of what would actually happen.
Makes sense. I was thinking to actually weigh in the cost of compilation but you're right, that makes no sense.  That cost is only paid at config validation time and isn't runtime relevant.

> Additionally the list of regexes at the top of the file should be real use cases to get a more realistic benchmark.
Fair point, I'll have to work on that. It's obviously easier to write code to generate random strings then to come up with hundreds of real IDs. I'll see what I can do.

In the meantime, I adjusted my simple benchmark to account for the first part of the feedback before getting to work on the real benchmark. Here's the output after adjusting to use `timeit` and removing the compilation time from the test (it's also running 10k times how hence the drastic difference in times):
```
One RE: 0.770606712
Many REs: 0.15945840999999994
```
Here's the new code for reference: https://paste.ubuntu.com/p/WxSQZK8tTs/

Reply:
So I ran the benchmarking suite before and after my changes.  Here's the output from my branch:
```
hass --script benchmark logbook_filtering_attributes
Using event loop: asyncio.unix_events
Benchmark logbook_filtering_attributes done in 0.21128311799999988s
Benchmark logbook_filtering_attributes done in 0.19500868800000015s
Benchmark logbook_filtering_attributes done in 0.224449098s
Benchmark logbook_filtering_attributes done in 0.21778510299999976s
Benchmark logbook_filtering_attributes done in 0.20733284900000015s
Benchmark logbook_filtering_attributes done in 0.19565272800000022s
Benchmark logbook_filtering_attributes done in 0.21827075699999998s
Benchmark logbook_filtering_attributes done in 0.19841470599999989s
Benchmark logbook_filtering_attributes done in 0.19276603700000017s
Benchmark logbook_filtering_attributes done in 0.197897223s
```
And here's the output from current `dev`:
```
hass --script benchmark logbook_filtering_attributes
Using event loop: asyncio.unix_events
Benchmark logbook_filtering_attributes done in 0.20831259699999993s
Benchmark logbook_filtering_attributes done in 0.20050133199999998s
Benchmark logbook_filtering_attributes done in 0.2051723080000003s
Benchmark logbook_filtering_attributes done in 0.21236684100000014s
Benchmark logbook_filtering_attributes done in 0.18955249500000004s
Benchmark logbook_filtering_attributes done in 0.18459170999999985s
Benchmark logbook_filtering_attributes done in 0.17939249299999993s
Benchmark logbook_filtering_attributes done in 0.18714338099999983s
Benchmark logbook_filtering_attributes done in 0.19031690199999973s
Benchmark logbook_filtering_attributes done in 0.19170428899999958s
```

There's not really a difference from what I can see. I'm somewhat confused though, I looked at the script and it doesn't actually seem to add any filters to test against. The line that creates the filter is just this: `entities_filter = logbook._generate_filter_from_config({})`. It also just runs a single state changed event over and over again, right?

To my knowledge, entity filter won't even put this through any filtering logic currently. Without any filters specified its just going to return a lambda that says always pass. So do you want me to add a new benchmark in here for this?

Reply:
I think something like this is a more realistic benchmark:
https://paste.ubuntu.com/p/wg2T9tSBB6/
```
One RE: 0.7973940710071474
Many REs (comp): 0.8852071089204401
Many REs: 0.1952527731191367
```

Reply:
Oh wow.  I didn't even consider that the `any` syntax could add so much overhead, that was silly of me. I wrote the perftest and then figured out how to simplify the code with `any` later. Good lesson in always perftesting the exact code.  Will definitely make that adjustment now.

Reply:
> To my knowledge, entity filter won't even put this through any filtering logic currently. Without any filters specified its just going to return a lambda that says always pass. So do you want me to add a new benchmark in here for this?

A new benchmark for filters would be great as they have been a bottleneck in the past

Reply:
So I took a stab at creating a benchmark test but I'm not sure how to actually run it. I called my new benchmark `logbook_filtering_entity_id` and added the `@benchmark` decorator. But when I try and do `hass --script benchmark logbook_filtering_entity_id` it tells me that's an invalid choice. 

Is there something else I have to update or a script to run to let the system know there's a new benchmark test available?

Reply:
That should work.  If you push up the change, I can take a look.

Reply:
Ok just pushed it. Note that I have not been able to run it so there may be bugs I have to iron out still. `pylint` passed but that only goes so far.

Reply:
```
root@ha-dev:~/home-assistant#  hass --script benchmark logbook_filtering_entity_id
Using event loop: asyncio.unix_events
Benchmark logbook_filtering_entity_id done in 0.6681695310398936s
Benchmark logbook_filtering_entity_id done in 0.7050089759286493s
Benchmark logbook_filtering_entity_id done in 0.6978548308834434s
Benchmark logbook_filtering_entity_id done in 0.7396661369130015s
Benchmark logbook_filtering_entity_id done in 0.7398890401236713s
Benchmark logbook_filtering_entity_id done in 0.7078370559029281s
```

Its working for me

Reply:
Oh I get it, I had to deactivate and reactivate my venv so the list of commands get reloaded.  Yea I see it running now too.